include/mxnet/ndarray.h

```diff
@@ -694,9 +694,13 @@ class NDArray {
   /*
    * Create NDArray from mkldnn memory.
    * mkldnn_mem The mkldnn memory to be managed.
-   * static_data If true, mkldnn memory won't be freed on destruction.
    */
-  explicit NDArray(const mkldnn::memory *mkldnn_mem, bool static_data = true);
+  explicit NDArray(const std::shared_ptr<mkldnn::memory> &mkldnn_mem);
```

Comment:
Is this not also a breaking API change?

Reply:
Yes, but it's designed as an internal API, and it's enabled for mkldnn build only. Can we seperate internal API from user API?

Reply:
While the c_api is generally an internal api I would say that ndarray.h is actually a user facing API for C++ users.

Reply:
For ndarray class, can we separate internal API with user facing API? Those class functions shouldn't be used for c++ users as they are backend specific APIs.